src/enforcers/SwapOfferEnforcer.sol

```diff
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: MIT AND Apache-2.0
+pragma solidity 0.8.23;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { ExecutionLib } from "@erc7579/lib/ExecutionLib.sol";
+import { ModeLib } from "@erc7579/lib/ModeLib.sol";
+
+import { CaveatEnforcer } from "./CaveatEnforcer.sol";
+import { ModeCode, Delegation } from "../utils/Types.sol";
+import { IDelegationManager } from "../interfaces/IDelegationManager.sol";
+
+/**
+ * @title SwapOfferEnforcer
+ * @dev This contract enforces a swap offer, allowing partial transfers if the order is not filled in a single transaction.
+ * @dev This caveat enforcer only works when the execution is in single mode.
+ * @dev The redeemer must include an allowance delegation when executing the swap to ensure payment.
+ */
+contract SwapOfferEnforcer is CaveatEnforcer {
+    using ExecutionLib for bytes;
+    using ModeLib for ModeCode;
+
+    struct SwapOffer {
+        address tokenIn;
+        address tokenOut;
+        uint256 amountIn;
+        uint256 amountOut;
+        uint256 amountInFilled;
+        uint256 amountOutFilled;
+        address recipient;
+    }
+
+    struct SwapOfferTerms {
+        address tokenIn;
+        address tokenOut;
+        uint256 amountIn;
+        uint256 amountOut;
+        address recipient;
+    }
+
+    struct SwapOfferArgs {
+        uint256 claimedAmount;
+        IDelegationManager delegationManager;
```

Comment:
It is safer not to allow the redeemer to provide the delegationManager address in the args, instead always use msg.sender as the other enforcers. As it is right now someone could call the beforeHook directly and modify values that he shouldn't.